mixins/provider-mixin.js

```diff
@@ -0,0 +1,38 @@
+export const ProviderMixin = superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._providers = new Map();
+		this.addEventListener('d2l-request-provider', this._handleRequest);
+	}
+
+	_handleRequest(e) {
+		if (this._providers.has(e.detail.key)) {
+			e.detail.provider = this._providers.get(e.detail.key);
+			e.stopPropagation();
+		}
+	}
+
+	provide(key, factory) {
+		this._providers.set(key, factory);
+	}
+
+	delete(key) {
+		this._providers.delete(key);
+	}
+};
+
+export const RequestProviderMixin = superclass => class extends superclass {
```

Comment:
This is more personal preference, but I think the names `ProviderMixin` and `RequesterMixin` were a bit more clear when I first saw them. The file could still be called `provider-mixin.js` since being a "provider" is still the main thing happening here, and the method here could still be `requestProvider` since that's what it's doing. 🤷‍♂

Reply:
The `requestProvider` method name doesn't seem quite right to me, since it's not requesting a provider... it is requesting a object from the provider.  `requestObject` seems too generic to me.  `requestDependency` might be better.  Open to other ideas too.  🤷‍♂ 

Somewhat separate thought, what happens in the case where none of the providers fulfill the request?  That might be an error case, but the consumer would have to check the return value to be safe.  We may be able to improve the ergonomics of this, either by returning a promise.
```
requestDependency(key) {
    const event = new CustomEvent('d2l-request-provider', {
        detail: { key },
        bubbles: true,
        composed: true,
        cancelable: true
    });

    this.dispatchEvent(event);
    if (event.detail.provider) return event.detail.provider;
    return Promise.reject();
}
```
The other thing I was thinking of... does this need to be a mixin?  It sees like it _could_ be a method that gets imported.  The only tie to the component that is needed is the DOM.  Food for thought... 
```
import { requestDependency } from '@brightspace-ui/core/mixins/provider-mixin.js';
...
// inside the component...
const logging = await requestDependency(this, 'some-logging-api');
```

Reply:
**When no providers fulfill the request:**
Good call. Although leaving async aside for now, it could just throw a `ProviderNotFound` exception?

Mixin vs. regular import:
If it's not a `LitElement`, `this.dispatchEvent` isn't going to work though ya? And then the `requestDependency` call would need to change to take an element from which to dispatch the event, which kind of filths up the API.

Reply:
Admittedly I'm mostly the messenger here, so I'll defer deeper analysis to @mdgbayly, but I think the idea behind `requestProvider` is similar to why the variable is called `factory` - we're not requesting the data directly, but rather are requesting a function that can be called which will return the data. That said, perhaps that's unnecessary and we can just return the data directly? (Via Promise does seem like a good call.)

As for mixin-vs-not, we're calling `dispatchEvent` in the requester and `addEventListener` in the provider, which I thought meant we needed to exist in mixin-land? But if not, I've probably made a few mixins that didn't need to be mixins 😅

Reply:
Yeah, as per Dave's comment, not making it a mixin does muddy up the API slightly, and is certainly a reason to go the mixin route.  It is arguably more reusable as just a method, but if we ever did need to dispatch from a non-Lit context, we could factor this method out of the mixin in the future.  

In the same vein, the async work I recently did relies on async elements dispatching `pending-state` - I considered exporting a method to share the construction of such promise-carrying event, but I have not yet.  I hadn't considered making a mixin to share this logic.  We could though, but I can't help but feel it's a bit excessive.

Not that this is the case, but in general I think we should be careful not to use mixins as a defacto method of sharing code.  We definitely created quite a few "Behaviors" for our Polymer components that could have been less opinionated, so we should be watchful for that.

Reply:
Yeh , i think the whole 'requestProvider` mixin was originally intended to signify that the client was `requesting a provider` so request is a verb rather than a noun, but again we should use whatever terminology we find most clear.